bd-client-common/src/file.rs

```diff
@@ -0,0 +1,41 @@
+// shared-core - bitdrift's common client/server libraries
+// Copyright Bitdrift, Inc. All rights reserved.
+//
+// Use of this source code is governed by a source available license that can be found in the
+// LICENSE file or at:
+// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt
+
+use crate::zlib::DEFAULT_MOBILE_ZLIB_COMPRESSION_LEVEL;
+use flate2::read::{ZlibDecoder, ZlibEncoder};
+use flate2::Compression;
+use std::io::Read;
+
+pub fn write_compressed_protobuf<T: protobuf::Message>(message: &T) -> Vec<u8> {
+  let bytes = message.write_to_bytes().unwrap();
+  let mut encoder = ZlibEncoder::new(
+    &bytes[..],
+    Compression::new(DEFAULT_MOBILE_ZLIB_COMPRESSION_LEVEL),
+  );
+  let mut compressed_bytes = Vec::new();
+  encoder.read_to_end(&mut compressed_bytes).unwrap();
+  compressed_bytes
+}
+
+pub fn read_compressed_protobuf<T: protobuf::Message>(
+  compressed_bytes: &[u8],
+) -> anyhow::Result<T> {
+  // We should never write empty files. If there is no data this was a partial write, full disk
+  // issue, or some other problem. We use zlib for compression which includes a CRC at the end
+  // so as long as the file is not empty we can be sure that the data is not corrupted.
+  if compressed_bytes.is_empty() {
+    anyhow::bail!("unexpected empty file");
+  }
+
+  // The files are likely not large enough to deal with streaming decompression on top of flate2.
+  // For now we just read the entire thing and then decompress it in memory. We can consider
+  // streaming later.
+  let mut decoder = ZlibDecoder::new(compressed_bytes);
```

Comment:
This is probably old code but can we pass a Reader to the decoder so that we don't have to pull the whole file contents into memory? 

Same above /w using write_to_writer instead of `write_to_bytes`

Reply:
The problem ends up being async vs not. We would need an adapter to use this code from async and I don't think it's worth it given the comment here about file sizes not being very large. I think we could fix this in the future if we ever deal with larger files.